packages/testlab/src/test-sandbox.ts

```diff
@@ -0,0 +1,59 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/testlab
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {tmpdir} from 'os';
+import {createHash} from 'crypto';
+import {resolve, join, parse} from 'path';
+import * as util from 'util';
+import {mkdtempSync, mkdir as mkdir_, copyFile as copyFile_} from 'fs';
+const promisify = util.promisify || require('util.promisify/implementation');
+
+const rimraf = promisify(require('rimraf'));
+const mkdir = promisify(mkdir_);
+const copyFile = promisify(copyFile_);
+
+export class TestSandbox {
+  private path: string;
+
+  constructor() {
+    this.create();
+  }
+
+  private create(): void {
+    this.path = mkdtempSync(join(tmpdir(), 'lb4-temp-'));
```

Comment:
1. Please use async version.

2. I personally prefer to place the sandbox directory in project's directory tree (typically in `test/sandbox`), because it makes it much easier to inspect the files seen by a test after the test failed. This has downsides when dealing with `npm install` in scaffolded packages (because the package sees dependencies of the tested packages and possibly our monorepo), but I think those situations should be exceptional and therefore handled as an exception, with the rest of our code falling back to using sanbox inside the project.

For example, the `TestSandbox` class can accept a parameter for the directory where to create the sandbox in.

```js
// sample usage in packages/core/cli/test/clone-example.js
const sandbox = new TestSanbox(path.resolve(__dirname, 'sandbox'));
```

Thoughts?

If we decide to keep creating random directories in the temp folder, then I am proposing to change `lb4-` prefix to `lb-`, so that we don't have to touch it when we eventually start working on version 5, 6, etc.

Reply:
Oh, so you cannot use the async version here because this code is called from the constructor. Makes sense.

Have you considered a private constructor and an async factory method?

```ts
export class TestSandbox {
  private constructor(private path: string) {}

  public async createRandom() {
    return new TestSandbox(await mkdtempSync(join(tmpdir(), 'lb4-temp-')));
  }

 public async resetStatic(dir: string) {
    await rimraf(dir);
    await mkdirp(dir);
    return new TestSandbox(dir);
  }

  // the rest of your API goes here
}
```

Reply:
Hmm, on the second thought, when considering how I would like to use this new sandbox helper in [packages/cli/test/clone-example.test.js](https://github.com/strongloop/loopback-next/blob/0df1e0d1affbadf9e9a3a4927048f179ee22763c/packages/cli/test/clone-example.test.js), I think my proposed API will offer poor ergonomics.

Ideally, I think I'd like to do something like the following:

```ts
// in packages/cli/test/helpers.ts
// create a TestSanbox instance using test/sandbox as the directory
// this should not necessarily create the directory yet!
export const sandbox = new TestSanbox(path.resolve(__dirname, 'sandbox'));

// in packages/cli/test/clone-example.test.js
describe('cloneExampleFromGitHub', () => {
  // ensure the sandbox directory exists (when the first test is run)
  // and is empty (does not contain files from previous tests)
  beforeEach(() => sandbox.reset());

  // individual tests follow
});
```

I have one more concern about creating sandbox directories in the temp folder. Who is going to delete those directories? If we delete the sandbox directory after the test finished, then it will be very difficult to examine the exact contents of sandbox files and run the failing code manually in the target environment. See [Clean the database before each test](http://loopback.io/doc/en/lb4/Testing-your-application.html#clean-the-database-before-each-test) in our "Testing your application" docs, it's a similar problem.